[Documentation] Squiz: Superfluous Whitespace

jonmcp opened the pull request:
## Description
This PR adds documentation for the `Squiz.WhiteSpace.SuperfluousWhitespace` sniff.

## Suggested changelog entry
Add documentation for the Squiz Superfluous Whitespace sniff

## Related issues/external references
Part of #148 

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix _(non-breaking change which fixes an issue)_
- [ ] New feature _(non-breaking change which adds functionality)_
- [ ] Breaking change _(fix or feature that would cause existing functionality to change)_
    - [ ] This change is only breaking for integrators, not for external standards or end-users.
- [x] Documentation improvement


## PR checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] I have checked there is no other PR open for the same change.
- [x] I have read the [Contribution Guidelines](https://github.com/PHPCSStandards/PHP_CodeSniffer/blob/master/.github/CONTRIBUTING.md).
- [x] I grant the project the right to include and distribute the code under the BSD-3-Clause license (and I have the right to grant these rights).
- [ ] I have added tests to cover my changes.
- [x] I have verified that the code complies with the projects coding standards.
- [ ] [Required for new sniffs] I have added XML documentation for the sniff.

jrfnl commented on src/Standards/Squiz/Docs/WhiteSpace/SuperfluousWhitespaceStandard.xml:
```diff
@@ -0,0 +1,66 @@
+        </code>
+        <code title="Invalid: Whitespace used before content in file.">
+        <![CDATA[
+        <em></em>
```
```suggestion
<em>        </em>
```

jrfnl commented on src/Standards/Squiz/Docs/WhiteSpace/SuperfluousWhitespaceStandard.xml:
```diff
@@ -0,0 +1,66 @@
+    ]]>
+    </standard>
+    <code_comparison>
+        <code title="Valid: No whitespace found at end of line.">
+        <![CDATA[
+echo 'hello';
+]]>
+        </code>
+        <code title="Invalid: Whitespace found at end of line.">
+        <![CDATA[
+echo 'hello';<em>   </em>
+]]>
+        </code>
```
When looking at these code sample in the command line, they don't look any different. That can't really be helped, but I wonder if it would clarify things if instead of "hello", the text used would describe what's happening ?

Something like "semicolon followed by new line char" vs "trailing spaces after semicolon" ?

jonmcp replied:
That makes a lot of sense. I've reworded all of the code examples in aa088bdfc

jrfnl commented on src/Standards/Squiz/Docs/WhiteSpace/SuperfluousWhitespaceStandard.xml:
```diff
@@ -0,0 +1,66 @@
+<documentation title="Superfluous Whitespace">
+    <standard>
+    <![CDATA[
+    There should be no extra whitespace at the start of a file.
```
Maybe ?

```suggestion
    There should be no superfluous whitespace at the start of a file.
```

jonmcp replied:
Corrected in aa088bdfc

jrfnl reviewed (commented):
@jonmcp Thank you for taking on this sniff !

Aside from the inline notes, I noticed that the "Additional whitespace found at end of file" error is not documented. Could you please add the documentation for that ?
It's basically about whitespace found after a PHP close tag (without inline HTML contents after).

Nitpick: it looks like the CDATA close tag for all code samples is not indented ?

jonmcp commented:
Thanks for the review, @jrfnl; I'll address your feedback today.

> it looks like the CDATA close tag for all code samples is not indented

I found that the existing documentation didn't seem to agree on whether the closing CDATA tag should be indented or not, so I chose to follow the format of most of the other `Squiz` sniffs . I should perhaps have asked prior to opening my PR, and am sorry for not doing so.  
I will of course address the indentation here. Would it be helpful if I were to open a separate PR to normalise the indentation of closing CDATA tags across the existing `Squiz` documentation to meet the style of the other standards? I could perhaps open a PR for each standard where there are issues?

jonmcp commented:
Thanks @jrfnl!

Once I've addressed feedback in my existing PRs, I'll prioritise normalising the existing sniff docs over creating new docs, so that it's easier going forward.

I think I've addressed all of your current feedback in this PR, but I'm not sure what the correct etiquette is now. When I think a PR is ready for a re-review, would it be helpful for me to modify the labels on the PR, or is it best if I leave those alone? I did re-read the contribution guidelines, and couldn't see anything specific about labels or whatnot, but please forgive me if I've missed something there. I'd like to be as helpful as I can be without getting in the way! 🙂

jrfnl commented:
> I think I've addressed all of your current feedback in this PR, but I'm not sure what the correct etiquette is now. When I think a PR is ready for a re-review, would it be helpful for me to modify the labels on the PR, or is it best if I leave those alone? I did re-read the contribution guidelines, and couldn't see anything specific about labels or whatnot, but please forgive me if I've missed something there. I'd like to be as helpful as I can be without getting in the way! 🙂

@jonmcp I don't think you can adjust the labels, unless you have triage/commit access to the repo... or am I missing something ?

If you would be able to update the labels, then removing the "Status: awaiting feedback" label would be the way to go. For PRs, I use that label to filter out PRs which have been reviewed and are waiting for an update/reply.

I still mean to look into creating a GH Actions workflow to automatically remove that label when a PR has been updated and/or received a reply from the original author, but I haven't got round to that yet (well, I did look into it a bit and found that it probably wouldn't be easy to create, so I left it for now)

Other than that, there's nothing for you to do, I have email notifications turned on for this repo, so I automatically get notified of updates and will look at them when I have time.

jrfnl commented on src/Standards/Squiz/Docs/WhiteSpace/SuperfluousWhitespaceStandard.xml:
```diff
@@ -0,0 +1,99 @@
+    <code_comparison>
+        <code title="Valid: No whitespace found at end of line.">
+        <![CDATA[
+echo 'semicolon followed by new linechar';
```
```suggestion
echo 'semicolon followed by new line char';
```

jonmcp replied:
should be fixed in fe18bcb78

jrfnl commented on src/Standards/Squiz/Docs/WhiteSpace/SuperfluousWhitespaceStandard.xml:
```diff
@@ -0,0 +1,99 @@
+    </code_comparison>
+    <standard>
+    <![CDATA[
+    There should be no superfluous whitespace at the end of a file.
+    A single final newline is allowed. A closing PHP tag is optional and ignored.
```
This description is incorrect and neither of the code samples would trigger the sniff.

As per my previous review:

> It's basically about whitespace found after a PHP close tag (without inline HTML contents after).

So that rule is not about whitespace between the last code and the close tag, but about whitespace _after a close tag when the close tag is not followed by inline HTML.

Does that help ?

jonmcp replied:
Apologies. I've clearly gotten confused somewhere. I'll try to be more careful in future. I've rewritten this in a584db264. Thanks for your patience.

jrfnl replied:
Happens to the best of us. Looks much better now.

jrfnl reviewed (changes requested):
@jonmcp Thanks for making that update!

We're close, but the new section still needs work. I've left two notes inline.

jonmcp commented:
> I don't think you can adjust the labels… Other than that, there's nothing for you to do…

You are indeed correct. I was hoping I might be able to help save a little time, but I ended up doing the opposite. Sorry about that, and thanks for your patience.

jrfnl reviewed (approved):
Thank you @jonmcp for working on this and making those updates!

It all looks good to me now, so the docs will be included in the next release.

For the record: on the command-line the code samples don't always show a difference. This is due to the sniff being about whitespace and cannot be helped.
The text strings used within the code samples stating what is happening, should still illustrate this well enough and are a good solution for the docs for this particular sniff.
